pyop2/codegen/rep2loopy.py

```diff
@@ -85,6 +91,210 @@ def petsc_function_lookup(target, identifier):
     return None
 
 
+class INVCallable(loopy.ScalarCallable):
+    def __init__(self, name, arg_id_to_dtype=None,
+                 arg_id_to_descr=None, name_in_target=None):
+
+        super(INVCallable, self).__init__(name,
+                                          arg_id_to_dtype=arg_id_to_dtype,
+                                          arg_id_to_descr=arg_id_to_descr)
+
+        self.name = name
+        self.name_in_target = name_in_target
+
+    def with_types(self, arg_id_to_dtype, kernel, callables_table):
+        for i in range(len(arg_id_to_dtype)):
+            if arg_id_to_dtype.get(i) is None:
+                # the types provided aren't mature enough to specialize the
+                # callable
+                return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                        callables_table)
+
+        mat_dtype = arg_id_to_dtype[0].numpy_dtype
+        name_in_target = "inverse"
+
+        return (self.copy(name_in_target=name_in_target,
+                          arg_id_to_dtype={-1: NumpyType(mat_dtype, target=loopy.CTarget()), 0: NumpyType(mat_dtype, target=loopy.CTarget()), 1: NumpyType(numpy.int32, target=loopy.CTarget())}),
+                callables_table)
+
+    def emit_call_insn(self, insn, target, expression_to_code_mapper):
+        assert self.is_ready_for_codegen()
+
+        assert isinstance(insn, loopy.CallInstruction)
+
+        parameters = insn.expression.parameters
+
+        parameters = list(parameters)
+        par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+        parameters.append(insn.assignees[-1])
+        par_dtypes.append(self.arg_id_to_dtype[0])
+
+        mat_descr = self.arg_id_to_descr[0]
+        arg_c_parameters = [
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            if isinstance(par, SubArrayRef) else
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            for par, par_dtype in zip(parameters, par_dtypes)
+        ]
+        c_parameters = []
+        c_parameters.insert(0, arg_c_parameters[-1])  # t1
+        c_parameters.insert(1, arg_c_parameters[0])  # t0
+        c_parameters.insert(2, numpy.int32(mat_descr.shape[0]))  # n
+        return var(self.name_in_target)(*c_parameters), False
+
+    def generate_preambles(self, target):
+        assert isinstance(target, CTarget)
+        inverse_preamble = """
+            #include <string.h>
+            #include <stdio.h>
+            #include <stdlib.h>
+            #ifndef Inverse_HPP
+            #define Inverse_HPP
```

Comment:
This is presumably to guard that this code is not emitted more than once. This seems ugly though, why does loopy not rationalise these generated preambles?

Reply:
Actually maybe they are. The first argument in the yield might be an identifier that prevents this.